packages/databyss-editor/state/reducer.ts

```diff
@@ -78,9 +86,103 @@ export default (
       const { payload } = action
 
       // default nextSelection to `payload.selection` (which may be undef)
-      const nextSelection = payload.selection
+      let nextSelection = payload?.selection
 
       switch (action.type) {
+        case CUT: {
+          deleteBlocksAtSelection({ state: state, draftState: draft })
+          break
+        }
+        case PASTE: {
+          const _frag = payload.data
+
+          // check if paste is occuring on an atomic block
+
+          const {anchor: _startAnchor} = sortSelection(state.selection)
+          const _startBlock = state.blocks[_startAnchor.index]
+
+          // if selection in the middle of atomic prevent paste
+          if(!isSelectionCollapsed(state.selection) || _startBlock.text.textValue.length !== _startAnchor.offset){            
+            if( isAtomicInlineType(_startBlock.type)){
+              break
+            }
+          }
+
+          if (!isSelectionCollapsed(state.selection)) {
+            deleteBlocksAtSelection({ state: state, draftState: draft })
+          }
+
+          if (_frag.length) {
+            // if fragment length is greater than 1 split blocks and insert the fragment
+            const _isCurrentBlockEmpty = !draft.blocks[
+              draft.selection.anchor.index
+            ].text.textValue.length
+
+            // splice blocks at current index
+            if (
+              _frag.length > 1 ||
+              _isCurrentBlockEmpty ||
+              isAtomicInlineType(_frag[0].type) ||
+              isAtomicInlineType(_startBlock.type)
+
+            ) {
+              const _spliceIndex = _isCurrentBlockEmpty
+                ? draft.selection.anchor.index
+                : draft.selection.anchor.index + 1
+
+              // insert block at index
+              draft.blocks.splice(
+                _spliceIndex,
+                _isCurrentBlockEmpty ? 1 : 0,
+                ..._frag
+              )
+
+              draft.operations.reloadAll = true
+
+              // set selection
+              const _selectionIndex = _spliceIndex + _frag.length - 1
+              const _offset =
+                draft.blocks[_selectionIndex].text.textValue.length
+
+              const _nextSelection = {
+                _id: draft.selection._id,
+                anchor: { index: _selectionIndex, offset: _offset },
+                focus: { index: _selectionIndex, offset: _offset },
+              }
+              nextSelection = _nextSelection
+            } else {
+              // merge fragment at current block
+              const { blocks, selection } = draft
+              const { anchor } = selection
+              const _index = anchor.index
+              const _mergedBlock = insertBlockAtIndex({
```

Comment:
Ah, ok, this function's purpose is clearer here. So a better signature might be 
```
insertText(
  { block, text, offset } : 
  { block: Block, text: Text, offset: number }
)
``` 
and you'd call it like
```
insertText({ 
  blocks: blocks[anchor.index], 
  text: frag[0].text, 
  offset: selection.anchor.offset,
})
```